docs/design/thee_datahall.md

```diff
@@ -0,0 +1,197 @@
+
+### Configure locality
+
+The `three_data_hall` deployment requires the `locality_data_hall` locality to be set and the `three_datacenter` requires the `locality_dcid` to be set.
+There must be at least and at most 3 different values for these localities.
+Additionaly to these localties at least two different `locality_zone` per `data_hall`/`datacenter` is required to fullfill the requirements.
+The `locality_zone` can be configured with [FoundationDBClusterFaultDomain](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterfaultdomain).
+In addition to that the user can configure additional variables that the sidecar will use for substition in the [FoundationDBClusterSpec](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterspec) with the `sidecarVariables` list.
+The current implementation would allow to set `dataHall: $AZ` and in `sidecarVariables` we would list `AZ` to define the `locality_data_hall` based on an environment variable that will be replaced by the sidecar with the actual value.
+Instead of having these different mechanisims spread across different settings I propose to have a new `localities` setting in the `FoundationDBClusterSpec`:
+
+```yaml
+localities:
+- key: "locality_data_hall"
+  value: ""
+  valueFromEnv: ""
+  valueFromNode: "topology.kubernetes.io/zone"
+  topologyKey: ""
+- key: "locality_zone"
+  value: ""
+  valueFromEnv: "MyFancyZone"
+  valueFromNode: ""
+  topologyKey: "kubernetes.io/hostname"
+```
+
+The `topologyKey` will be used for `topologySpreadConstraints` and for `PodAntiAffinity` and only has to be set if `valueFromNode` is empty, otherwise it will default to that value.
+The `FoundationDBClusterFaultDomain` would then be reduced to only have `key`, `zoneCount` and `zoneIndex` since the locality will be configured in `localities`.
+For all `localities` that define a `valueFromEnv` we would add the key to the `--substitute-variable` flag.
+For `valueFromNode` we have to modify the sidecar to allow it to read labels from Kubernetes nodes (see [Related Links](#related-links)) and pass that information to the according new flag.
+This change should provide the most flexibility to the user to define the required/wanted localities.
+We would set `locality_instance_id` and `locality_machineid` to the current defaults but also allow the user to define custom localities.
+After that change we should deprecate `cluster.Spec.DataCenter` and `cluster.Spec.DataHall`.
+
+### Deployment model
+
+#### Single regional Kubernetes cluster
+
+The assumption here is that the user has a regional Kubernetes cluster that spreads across at least 3 different AZs.
+To ensure that all Pods are spread evenly across the AZs the operator should use [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints).
```

Comment:
Does the operator need to ensure that pods remain balanced after removing/excluding pods? 
E.g. after scaling down the cluster.

Reply:
I think that would be a useful feature. During a replacement everything should be fine we could only run in an issue when we do a shrink but the [chooseDistributedProcesses](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/controllers/cluster_controller.go#L1198) should take care of that. I will add some additional information in the design and we should add enough test cases to get confidence.

We also have an open issue for automatic balancing: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/499.

Reply:
Do we want to include topology spread constraints here, or use anti-affinities in this context and make topology spread constraints a follow up? Do we have concerns about older versions of Kubernetes where topology spread constraints may not be available?

Reply:
If we're going to move to topology spread constraints, how will we control the rollout of that change?

Reply:
I think the topology spread makes more sense in this specific use case. Topology spread is supported (stable) since `1.19` and beta in `1.18` (auto enabled) so I think using topology spread and increasing the required version of Kubernetes should be fine? Currently `1.19` is the oldest "supported" version of Kubernetes: https://kubernetes.io/releases/

Reply:
The rollout would require a replacement for stateful Pods (storage, log) and for stateless Pods we could do a recreate. For simplicity we might just want to replace all Pods.